src/libstd/keyword_docs.rs

```diff
@@ -1578,12 +1578,181 @@ mod type_keyword {}
 
 #[doc(keyword = "unsafe")]
 //
-/// Code or interfaces whose [memory safety] cannot be verified by the type system.
+/// Code or interfaces whose [memory safety] cannot be verified by the type
+/// system.
 ///
-/// The documentation for this keyword is [not yet complete]. Pull requests welcome!
+/// The `unsafe` keyword has two uses: to declare the existence of contracts the
+/// compiler can't check (`unsafe fn` and `unsafe trait`), and to declare that a
+/// programmer has checked that these contracts have been upheld (`unsafe {}`
+/// and `unsafe impl`, but also `unsafe fn` -- see below). They are not mutually
+/// exclusive, as can be seen in `unsafe fn`.
+///
+/// Unsafe Rust adds some capabilities to Safe Rust:
+///
+/// - Dereference [raw pointers]
+/// - Call `unsafe` functions (including foreign functions, some compiler
+/// intrinsics and the raw allocator)
+/// - Implement `unsafe` [`trait`]s
+/// - Mutate [`static`]s (including [`extern`]al ones)
+/// - Access fields of [`union`]s
+///
+/// # Why do we need a keyword for this ?
+///
+/// **No matter what, Safe Rust can't cause Undefined Behavior**. This is
+/// referenced to as [soundness]: a well-typed program actually has the desired
+/// properties. In Safe Rust, the language and the compiler do this mostly by
+/// themselves whereas in Unsafe Rust the programmer must help. The [Nomicon]
+/// has a more detailed explanation on the subject.
+///
+/// # The different meanings of `unsafe`
+///
+/// Depending on the place where it is used, the `unsafe` keyword meaning can
+/// change. See the following [discussion on Rust Internals] for more in-depth
+/// explanations.
+///
+/// Summarizing the main points:
+///
+/// - `unsafe fn`: calling this function means abiding by a contract the
+/// compiler cannot enforce.
+/// - `unsafe trait Trait`: implementing this [`trait`] means abiding by a
+/// contract the compiler cannot enforce.
+/// - `unsafe {}`: the contract necessary to call the operations inside the
+/// block has been checked and is guaranteed to be respected.
+/// - `unsafe impl`: the contract necessary to implement the trait has been
+/// checked and is guaranteed to be respected.
+///
+/// As of now (Rust 1.44), `unsafe fn` also acts like an `unsafe {}` block
+/// around the code inside the function. This means it is not just a signal to
+/// the caller, but also promises that the preconditions for the operations
+/// inside the function are upheld.
+///
+/// Mixing these two meanings can be confusing and [proposal]s exist to use
```

Comment:
```suggestion
/// inside the function are upheld.
/// Mixing these two meanings can be confusing and [proposal]s exist to use
```
This is logically one line of reasoning, IMO it should be one paragraph.

Reply:
Done !